sections/generic-best-practices/test-code-with-debugging-in-mind.md

```diff
@@ -0,0 +1,345 @@
+
+
+![The Cypress Test Runner showing the value of a logged object](../../assets/images/test-code-with-debugging-in-mind/cy-log-console.jpg)
+
+
+
+Storybook and Playwright already have the concept of `step` utilities that allow explaining in English what is happening in the test. Cypress does not have the same option, so the `cy.log` I propose is valuable, in my opinion.
+
+### Use clear selectors
+
+Look at this code
+
+```ts
+cy.get('textarea')
+  .eq(0)
+  .type(`{enter}{uparrow}${statements.createMutationGQLQuery}`, {
+    force: true,
+  });
+```
+
+What is `cy.get('textarea').eq(0)`? In the absence of better selectors, I suggest hiding them under Cypress aliases, such as
+
+```ts
+// Assign an alias to the most unclear selectors for future references
+cy.get('textarea').eq(0).as('actionDefinitionTextarea');
+cy.get('textarea').eq(1).as('typeConfigurationTextarea');
+```
+
+and then referring to them this way
+
+```ts
+cy.get('@actionDefinitionTextarea').clearConsoleTextarea().type(/* ... */);
+```
+
+to improve the readers' life.
+
+### Reducing data-testid attributes
+
+I do not want to speak about the value for the tests themselves and for their confidence, but only about the effect of data-testid attributes over the debugging phase.
+
+If an element with a data-tesid cannot be retrieved from the page, the possible problems are
+
+1.  The element is not there.
+2.  The element is there, but it has not the attribute.
+3.  The element is there, and it has the attribute but not the expected value.
+
+All the above problems **cause the developer to re-launch the tests, inspect the elements**, look for the test-related attributes, etc. Instead, if the tests are based on the textual contents, a screenshot is enough to understand if the text searched by the test is not there or if it is wrong.
+
+Also, some more cons for the engineers that have to deal with data-testid's
+
+1.  Test-related attributes must be maintained during refactors, but it is not easy when you have hundreds of them.
+2.  Test-related attributes are helpful if they are unique on the page. Another thing that is not easy to guarantee when you have hundreds of them
+
+My suggestion is to use data-testid attributes only for:
+
+-  Sections, not elements (ex. the Header, the Footer, , etc.) that allows reducing the scope of text-based searches. Here is an example
+
+```ts
+cy.get('[data-test="Actions list"]').within(() => { // <-- reduce the scope
+  cy.contains('login') // <-- the "login" text could exist more times in the page
+})
+```
+
+-  Non-text-based elements: icons, images, etc.
+
+Last but not least: I suggest valuing them with user-like values, not programmer-like ones (ex. "Actions List", not "actionsList"), especially when the section shows that exact text. This allows a direct connection between the code of the test, the Cypress' Test Runner, and the page's text content.
```

Comment:
Cypress guides say custom selectors  > text. Thoughts?

Reply:
> Cypress guides say custom selectors > text. Thoughts?

I do not agree, but it's a subjective topic, obviously 😊 also, attributes must be then maintained and some time is a nightmare in a real-world app. What's your experience here?

Reply:
In the CCTDD book app https://github.com/muratkeremozcan/tour-of-heroes-react-cypress-ts and in the real world at a large scale (common component libraries and apps with 500+ tests) the custom selectors worked great. 

I really like what can be done with JSX, template literals and custom selectors.